### PR TITLE
[doc] Add shape on undo redo

### PR DESCRIPTION
**Pull request by mcharfadi**
# Pull request template

## General purpose
What is the main goal of this pull request?
- [ ] Bug fixes
- [ ] New features
- [ ] Documentation
- [ ] Cleanup
- [ ] Tests
- [ ] Build / releng

## Project management
- [ ] Has the pull request been added to the relevant project and milestone? (Only if you know that your work is part of a specific iteration such as the current one)
- [ ] Have the `priority:` and `pr:` labels been added to the pull request? (In case of doubt, start with the labels `priority: low` and `pr: to review later`)
- [ ] Have the relevant issues been added to the pull request?
- [ ] Have the relevant labels been added to the issues? (`area:`, `difficulty:`, `type:`)
- [ ] Have the relevant issues been added to the same project and milestone as the pull request?
- [ ] Has the `CHANGELOG.adoc` been updated to reference the relevant issues?
- [ ] Have the relevant API breaks been described in the `CHANGELOG.adoc`? (Including changes in the GraphQL API)
- [ ] In case of a change with a visual impact, are there any screenshots in the `CHANGELOG.adoc`? For example in `doc/screenshots/2022.5.0-my-new-feature.png`

## Architectural decision records (ADR)
- [ ] Does the title of the commit contributing the ADR start with `[doc]`?
- [ ] Are the ADRs mentioned in the relevant section of the  `CHANGELOG.adoc`?

## Dependencies
- [ ] Are the new / upgraded dependencies mentioned in the relevant section of the `CHANGELOG.adoc`?
- [ ] Are the new dependencies justified in the `CHANGELOG.adoc`?

## Frontend

This section is not relevant if your contribution does not come with changes to the frontend.

### General purpose
- [ ] Is the code properly tested? (Plain old JavaScript tests for business code and tests based on React Testing Library for the components)

### Typing
We need to improve the typing of our code, as such, we require every contribution to come with proper TypeScript typing for both changes contributing new files and those modifying existing files.
Please ensure that the following statements are true for each file created or modified (this may require you to improve code outside of your contribution).

- [ ] Variables have a proper type
- [ ] Functions’ arguments have a proper type
- [ ] Functions’ return type are specified
- Hooks are properly typed:
	- [ ] `useMutation<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useQuery<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useSubscription<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useMachine<CONTEXT_TYPE, EVENTS_TYPE>(…)`
	- [ ] `useState<STATE_TYPE>(…)`
- [ ] All components have a proper typing for their props
- [ ] No useless optional chaining with `?.` (if the GraphQL API specifies that a field cannot be `null`, do not treat it has potentially `null` for example)
- [ ] Nullable values have a proper type (for example `let diagram: Diagram | null = null;`)

## Backend

This section is not relevant if your contribution does not come with changes to the backend.

### General purpose
- [ ] Are all the event handlers tested?
- [ ] Are the event processor tested?
- [ ] Is the business code (services) tested?
- [ ] Are diagram layout changes tested?

### Architecture
- [ ] Are data structure classes properly separated from behavioral classes?
- [ ] Are all the relevant fields final?
- [ ] Is any data structure mutable? If so, please write a comment indicating why
- [ ] Are behavioral classes either stateless or side effect free?

### How to test this PR?
_Please describe here the various use cases to test this pull request_

- [ ] Has the Kiwi TCMS test suite been updated with tests for this contribution?

### CODE REVIEW

File: doc/adrs/158_add_undo_redo_support.adoc [hunk 1]
```diff
@@ -0,0 +1,127 @@
+= ADR-158 Add undo redo support
+
+== Context
+
+Undo redo is not implemented on representations.
```
Comment (pcdavid) [inline doc/adrs/158_add_undo_redo_support.adoc]: This ADR only mentions representations and purely frontend state.
The real state of a project is on the backend (even in the DB), representations are only partial, ambiguous and opaque projections of that state.

I can currently define a tool which applies to a node on a diagram, makes changes on the underlying semantic data on the backend, but produces zero visible change on the frontend. A purely frontend approach to undo/redo can not handle this.

Comment (mcharfadi) [reply doc/adrs/158_add_undo_redo_support.adoc]: After more discussions, I completly changed the approach

File: doc/iterations/2024.11/add_undo_redo_support.adoc [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+- *Expected*: the element is renamed back to "Paul" and all elements that were impacted by the change also revert back to their previous state.
+- An user press ctrl+y
+- *Expected*: the element is renamed back to "Bernard" and all elements that were impacted by the change also revert back to their previous state.
+
+
+2. Several users are working on the same representation
+- An user u1 renames an element from "Paul" to "Bernard"
+- An user u2 delete the element
+- If An user u1 press ctrl+z
+- *Expected*: the semantic elements are reverted back to when the element was "Paul"
```
Comment (sbegaudeau) [inline doc/iterations/2024.11/add_undo_redo_support.adoc]: It would just undo the change not revert everything back to the previous state so in this case nothing would change.

**Review by sbegaudeau — COMMENTED** [review]
I'll perform the small change that I've requested and rebase the PR to merge it

**Review by sbegaudeau — APPROVED** [review]
